ext/TensorKitChainRulesCoreExt/constructors.jl

```diff
@@ -12,6 +12,17 @@ function ChainRulesCore.rrule(::Type{<:TensorMap}, d::DenseArray, args...; kwarg
     return TensorMap(d, args...; kwargs...), TensorMap_pullback
 end
 
+function ChainRulesCore.rrule(::Type{<:DiagonalTensorMap}, d::DenseVector, args...;
+                              kwargs...)
+    D = DiagonalTensorMap(d, args...; kwargs...)
+    project_D = ProjectTo(D)
+    function DiagonalTensorMap_pullback(Δt)
+        ∂d = project_D(unthunk(Δt)).data
+        return NoTangent(), ∂d, ntuple(_ -> NoTangent(), length(args))...
+    end
```

Comment:
This implementation slightly surprises me, I would have expected the projection to be based off `d`. In the end, these things probably boil down to the same thing?

```suggestion
    project_d = ProjectTo(d)
    function DiagonalTensorMap_pullback(Δt)
        ∂d = project_d(unthunk(Δt).data)
        return NoTangent(), ∂d, ntuple(_ -> NoTangent(), length(args))...
    end
    D = DiagonalTensorMap(d, args...; kwargs...)
```

Reply:
It seems to me this is not the same. The issue here, is that sometimes \Delta t may be of some non-diagonal type. I expect that in this situation your version will return incorrect tangent.  Am I missing something?

Reply:
I am also undecided between the two options. I agree that in most cases, e.g. tensor contractions, there should already have been a `ProjectTo(D)` being called on the adjoint variable `Δt` before it enters this `rrule`, and therefore not explicitly having `project_d(Δt)` in this rule might serve as a way to find missing projectors elsewhere.

On the other hand, from a user perspective, I can also see the advantage of just having this in `project_D` in here for safety. 

While in principle that is independent of still having to call `project_d = ProjectTo(d)` on the output `∂d`, it is probably true that `project_d` will not be doing anything (will act as the identity) if we already have projected `Δt` to a diagonal tensor with the right scalar type and storage type.